### PR TITLE
Implement support for masked stencil clears on Vulkan

### LINKED ISSUE #5585: Read Dead Redemption Bushes, Trees and Vegetation not rendering correctly [Bug]
### Description of the issue

When running Red Dead Redemption version 1.0.1 all the vegetation is not rendered correctly, causing a really psychedelic effect and consequently causing more visual glitches through the whole game. 

Such effect is already right at the first cinematic of the game and gets worse you progress through the story.
 

### Reproduction steps

Start the game and start the story mode.

### Log file

[Ryujinx_ 1.1.0-macos1_2023-08-16_19-22-30.log](https://github.com/Ryujinx/Ryujinx/files/12371172/Ryujinx_.1.1.0-macos1_2023-08-16_19-22-30.log)
[Ryujinx_ 1.1.0-macos1_2023-08-17_12-01-58.log](https://github.com/Ryujinx/Ryujinx/files/12371174/Ryujinx_.1.1.0-macos1_2023-08-17_12-01-58.log)
[Ryujinx_ 1.1.0-macos1_2023-08-17_12-29-57.log](https://github.com/Ryujinx/Ryujinx/files/12371175/Ryujinx_.1.1.0-macos1_2023-08-17_12-29-57.log)

### OS

Mac OS 14 Sonoma Beta 5

### Ryujinx version

1.1.0

### Game version

1.0.1

### CPU

M1

### GPU

M1

### RAM

8gb

### List of applied mods

No mods applied.

### Additional context?

_No response_

### PR DESCRIPTION
**Pull request by gdkchan**
Vulkan does not allow a stencil write mask to be specified for stencil clears, which means we can either "clear" the entire stencil value, or not clear it using `CmdClearAttachments`. This was being ignored before, and there was a TODO in the code to implement support for it. This change implements a fallback that is used for depth-stencil clears if the stencil mask is not all zeros or all ones, it performs the clear using a regular draw operation. This is the same thing we do for color clear with a component mask that is not RGBA.

Fixes afterimages on Red Dead Redemption, caused by depth tests failing.
Before:
image
After:
image
Fixes #5585.
Testing is welcome.

### CODE REVIEW

File: src/Ryujinx.Graphics.Vulkan/HelperShader.cs [hunk 1]
```diff
@@ -100,6 +101,12 @@ public HelperShader(VulkanRenderer gd, Device device)
             }, colorClearResourceLayout);
 
             _programColorClearUI = gd.CreateProgramWithMinimalLayout(new[]
+            {
+                new ShaderSource(ReadSpirv("ColorClearVertex.spv"), ShaderStage.Vertex, TargetLanguage.Spirv),
+                new ShaderSource(ReadSpirv("DepthStencilClearFragment.spv"), ShaderStage.Fragment, TargetLanguage.Spirv),
```
Comment (riperiperi) [inline src/Ryujinx.Graphics.Vulkan/HelperShader.cs:106]: Is this correct? The variable is `_programColorClearUI` but it's loading `DepthStencilClearFragment.spv`. The assignment below is `_programDepthStencilClear` but loads `ColorClearUIFragment.spv`.

**Comment by DiegoTegani** [comment]
I just took the test now using it without a filter and with FSR, everything is fine for me.

Specs:

Intel Core i7-7700 (I770)
NVIDIA GeForce GTX 1060
16GB RAM
NVIDIA Driver - Version 536.99

https://github.com/Ryujinx/Ryujinx/assets/45742711/9e11550e-2f88-4da3-acbc-5d0b8288b710

File: src/Ryujinx.Graphics.Vulkan/HelperShader.cs (hunk 1 above)
Comment (gdkchan) [reply src/Ryujinx.Graphics.Vulkan/HelperShader.cs]: Good catch, I messed that up somehow. Should be fixed now.

**Review by riperiperi — APPROVED** [review]
Awesome work.

**Comment by Defaalt** [comment]
I have been playing for at least an hour and I can confirm that this works perfectly. The glitching is completely gone !
Insane work ! thank you !

**Comment by solarmystic** [comment]
RDR is practically Playable with this PR. No more foliage smudging, thanks for the quick fix! 

image

**Review by AcK77 — APPROVED** [review]

**Comment by Emertels** [comment]
> RDR is practically Playable with this PR. No more foliage smudging, thanks for the quick fix! 
> 
> image
> 

Hello, friend! I tried to download this version in every possible way, but I couldn't. Can you host it on Google Drive so I can download it?

**Comment by ekuland** [comment]
> > RDR is practically Playable with this PR. No more foliage smudging, thanks for the quick fix!
> > image
> 
> Hello, friend! I tried to download this version in every possible way, but I couldn't. Can you host it on Google Drive so I can download it?

This change has been merged into the main build of the emulator.

**Comment by Emertels** [comment]
Thanks, friend! But these new versions don't seem to have texture support.

**Pull request merged**